Data dictionary valid values

orangejenny opened the pull request:
## Product Description
Spec: "External Release 2" in the [spec's Release Sequence](https://docs.google.com/document/d/1BPhdopyt07xNVv1uZ000094j25otKm6Deut1yXRSsyo/edit#heading=h.tlpffswyn3ey)

Showing valid values in the data dictionary:
![Screen Shot 2021-11-24 at 1 27 26 PM](https://user-images.githubusercontent.com/1486591/143296698-c1e57bb1-f1ac-4308-8cea-aeee854fac01.png)

Showing valid values while importing cases:
![Screen Shot 2021-11-24 at 1 40 43 PM](https://user-images.githubusercontent.com/1486591/143296734-728f97f7-32f3-431e-8f29-e1f760a15682.png)

Showing warnings after import:
![Screen Shot 2021-11-24 at 1 28 04 PM](https://user-images.githubusercontent.com/1486591/143296791-3c6a777f-31f2-441d-bf22-c14f87dfc277.png)

## Feature Flag
Validate data per data dictionary definitions during case import

## Safety Assurance

### Safety story
This has tests and has gone through QA. Issues should be limited to flagged domains and to the affected features (data dictionary & case import).

### Automated test coverage

There's decent test coverage in this PR.

### QA Plan

[QA ticket](https://dimagi-dev.atlassian.net/browse/QA-3436)

### Migrations

The migration in this PR just adds a new model.

- [x] The migrations in this code can be safely applied first independently of the code

### Rollback instructions

- [x] This PR can be reverted after deploy with no further considerations

### Labels & Review
- [x] Risk label is set correctly
- [x] The set of people pinged as reviewers is appropriate for the level of risk of the change

kmtracey committed:
Beginnings of UI for specifying allowed values...

...for select data types in data dictionary.

- add model for storing the allowed values and their descriptions
- fetch existing allowed values when initial page data is fetched
- display allowed values and an edit button when appropriate
- next step: make edit button functional

kmtracey committed:
Address initial review comments: try to be a bit more idiomatic/concise, and move html into template where it belongs.

kmtracey committed:
Merge branch 'master' into allowed_values_ui

kmtracey committed:
Use a uiElementKeyValueList to make the allowed values lists editable.

Additionally, change some text on the ui element (e.g. heading, add button) to be more user-friendly (reduce emphasis on "mapping"). Possibly these text bits should be made caller-configurable if these changes are not appropriate for all the places this widget is currently used.

Further, make the placeholders in the "mapping" fields caller-configurable.

Still TODO: enable "Save" button when the allowed values are edited, tests

kmtracey committed:
Merge branch 'master' into allowed_values_ui

kmtracey committed:
Enable Save button on allowed value modal "Done" click.

kmtracey committed:
Swith to the `on('change', ...)` made available via...

the `KeyValList` call to `hqMain.eventize` to enable the Save button, rather than listening to the Done button click.

This will fire regardless of how the modal is closed (x in top or Done button), which seems correct since the top X still records whatever changes have been made in what is subsequently shown to the user.

Also make the allowed values modal work properly for a new dynamically added property created after page load.

kmtracey committed:
Add the basics of export/import of allowed values in data dictionary.

kmtracey committed:
Address PR feedback and change to "valid values"...

...for user-visible references to what we have been calling allowed values. (Much of the code still uses the "alloewd value" terminology, if wanted I can work to change the internals as well, but that's a bigger task than time probably allows today.)

kmtracey committed:
Merge branch 'master' into allowed_values_ui

kmtracey committed:
Merge branch 'master' into allowed_values_ui

kmtracey committed:
Changes to account for new SQL model.

- include new model in domain export code
- ensure new model is deleted when domain is deleted

kmtracey committed:
Give user specific feedback on save of a too-long valid value. Refs QA-3469

kmtracey committed:
Beef up error catching/reporting on data dictionary upload/import.

Catch when a case type is missing valid values, and let the user know instead of producing a 500 error
Notice when a case property cell is empty, and let the user know
Notice when a property is listed as having valid values but not found in the base case type information, and let user know.

Also added some basic unit tests of the data dictionary export/import. Import tests for these error situations, export is a bit more basic (so far just tests the right sheets are in the exported file), but it should be easy enough to add on as the data dictionary is worked on more.

Refs QA-3457

kmtracey committed:
Validate select (multiple choice) values.

Extend the existing validity checking (had been added just for dates) to also include validing that select values match values provided for the property in the data dictionary. As with dates, do this if and only if the feature flag for data dictionary import validation is enabled.

kmtracey committed:
Couple of simplifications from review comments.

orangejenny committed:
Merge pull request #30251 from caktus/allowed_values_ui

Allowed (valid) values UI

kmtracey committed:
Fix typo in exception title.

orangejenny committed:
Merge pull request #30443 from caktus/validate-select-values

Validate select (multiple choice) values.

kmtracey committed:
Key/value widget updates for valid values editing

More specific/descriptive modal title (have caller of the shared code specify full title)
Allow caller to specify "subtitle" to show below the modal title
Allow caller to specify a maximum number of items to display for the display view

kmtracey committed:
Data dictionary UI updates.

Change user-visible name of "select" data type to "Multiple choice"
Switch from buttons with text to icons for deprecate/restore properties and FHIR paths
Swap positioning of FHIR resource property path and Valid Values column to keep icons apart
Add a tooltip for Valid Values & Formats column.

kmtracey committed:
Make new user-visible text "Multiple Choice" (consistent with "Phone Number" in same list)

kmtracey committed:
Switch page content to use full screen width.

kmtracey committed:
Add missing semicolons;

orangejenny committed:
Merge pull request #30456 from caktus/data-dictionary-ui-updates

Data dictionary UI updates.

orangejenny committed:
Merge pull request #30451 from caktus/keyval-widget-updates

Key/value widget updates for valid values editing

kmtracey committed:
Case importer matching page updates for valid values

Add "Valid values & formats" column with tooltip
Populate that column when possible (chosen property is a date or has valid values in data dictionary)
Move label text of checkbox for "create new property" to the header
Move "instead" from this label to the help text of "Did you mean..."

kmtracey committed:
Add missing all() on a related manager.

kmtracey committed:
Now that the date format is shown in Valid Values etc. column, remove it from under the data type select box.

orangejenny committed:
Merge pull request #30465 from caktus/data-dictionary-ui-updates

Now that the date format is shown in Valid Values etc. column, remove…

kmtracey committed:
Updates to error reporting on case import

Include a count of rows with invalid data, for each affected column name.
Include a sample of the invalid data.
For dates, include the expected format.
For multiple choice (select), include the configured valid values.

kmtracey committed:
Additional changes from basecamp feedback

Remove green coloring of "success" messages.
Change "Affected rows" to simply "Row(s)"
Allow any values for select types with no configured values in data dictionary.

kmtracey committed:
Shorten the data dictionary valid values sheet suffix.

To "-vl", allowing for Excel editing of the exported data dictionary for case types with names up to 28 chars in length. (Excel has a 31 character limit on sheet names).

Refs QA-3457

orangejenny committed:
Merge pull request #30508 from caktus/shorten-valid-values-sheet-suffix

Shorten the data dictionary valid values sheet suffix.

kmtracey committed:
Updates base on PR feedback, add python tests.

kmtracey committed:
Update from code review.

orangejenny committed:
Merge branch 'master' into jls/caktus-data-dictionary-release-2

orangejenny committed:
Merge pull request #30492 from caktus/clean-case-import-error-reporting-updates

Updates to error reporting on case import

orangejenny committed:
Merge pull request #30461 from caktus/case-import-matching-page-updates

Case importer matching page updates for valid values

orangejenny committed:
Merge branch 'master' into jls/caktus-data-dictionary-release-2

kmtracey committed:
Fix a couple of errors in data dictionary upload, ref QA-3594

First, a pure numeric value in an "allowed_value" cell would cause a 500 error. Fix by ensuring all values are processed as strings even if they start out numeric. Adjusted data and test for existing "clean" import test to unit test this case.

Second, rows that are "short" on cells would cause a cryptic error message. Though I'm not sure how variable-length rows are saved with Excel, the file "data_dictionary (28) (1).xlsx" attached to QA-3457 has them. On reflection, it seems fine to assume "blank" for entirely missing cells so I changed the code to assume blank rather than generate an error.

This 2nd one I wasn't able to add unit test for. I don't know how to generate an xlsx file with variable-length rows, and the utility I've seen used in case importer tests ("make_worksheet" from corehq.util.workbook_reading.adapters.raw_data) explicitly disallows variable-length rows. Let me know if a test is wanted for this and I'll figure something out but thought I should check before spending too much time on it.

orangejenny committed:
Merge pull request #30551 from caktus/qa-3594--improve-data-dictionary-upload

Fix a couple of errors in data dictionary upload, ref QA-3594

kmtracey committed:
Adjust the "reset to default" button as requested in figma.

orangejenny committed:
Merge pull request #30556 from caktus/matching-page-button-adjustment

Adjust the "reset to default" button as requested in figma.

kmtracey committed:
Improve data dictionary usablity

Adds title attribute for all buttons that previously had text and now have icons instead.
Titles have been set to be what the button text used to be.
Make these actual buttons, not <i> elements. Add styling rules to remove auto-button-appearance.
Tighten space between allowed values and "Edit" button (20px to 8px bottom margin).
Update copy on help text for the Valid Values & Formats column.
Force wrapping of very long case names in the sidebar.

kmtracey committed:
Move new css to wrap sidebar content to have global scope.

orangejenny committed:
Merge pull request #30608 from caktus/data-dictionary-usability-improvements

Improve data dictionary usablity

kmtracey committed:
During case import with mirrored domains enabled,

add a note to the tooltip for the valid values and formats column that indicates valid values may differ for mirrored domain.

Refs QA-3630

kmtracey committed:
Remove "mirrored" terminology from user-facing message.

orangejenny committed:
Merge pull request #30653 from caktus/qa-3630--add-mirroring-note

During case import with mirrored domains enabled,

kmtracey committed:
Show italic "blank" for empty valid values. Refs QA-3628

orangejenny committed:
Merge pull request #30660 from caktus/qa-3628--display-blank-values

Show italic "blank" for empty valid values. Refs QA-3628

snopoke committed:
Merge branch 'master' into jls/caktus-data-dictionary-release-2

kmtracey committed:
Refs QA-3628 Consistent use of italic blank for empty strings and wrap all-whitespace in quotes.

(This is in the display (non-edit view) of the key/values for the key/value widget.)

kmtracey committed:
Use DOMPurify to sanitize key/val for display.

(Also adjust formatting per sticklerbot, to some extent...not to the extent of changing variable name patterns in this code.)

orangejenny committed:
Merge pull request #30695 from caktus/qa-3628--whitespace-blank-handling

Refs QA-3628 Consistent use of italic blank for empty strings...

kmtracey committed:
Refs QA-3629 Refine behavior of matching excel fields to deprecated properties

Do not list deprecated properties in dropdown, but allow them to be typed and auto-suggested
When a deprecated property has been typed/suggested, show a warning that it is a deprecated property
Also show any formats/valid values in this case
Uncheck the "Create new property" in this case as well, since the property exists

In addition, this PR sanitizes allowed values input into the data dictionary before sending to the backend.
(Having script tags (for example) in the backend for suggested valid values causes the importer matching page
to not work as expected -- it seems the js initial data doesn't get deserialied properly -- so stripping in
the same way as is done for display on data dictiony seemed easist to fix this oddity.)

Finally, as requested by Daniel, disable the Next button on the initial upload page when there is no file
chosen.

kmtracey committed:
Refs QA-3629 js fixes per sticklerbot

kmtracey committed:
Refs QA-3629 use friendly-named utils for button enable/disable,

also limit return of deprecated properties by utility function to code requesting deprecated properties.

orangejenny committed:
Merge pull request #30760 from caktus/QA-3629--refine-behavior-matching-deprecated-properties

Refs QA-3629 Refine behavior of matching excel fields to deprecated p…

kmtracey committed:
Refs QA-3629 some final tweaks to case importer validation

Reverse the order of the warning messages around suggested/deprecated properties
Suppress the the appearance of the Valid Values and Formats column entirely when the toggle to validate data is not enabled.

orangejenny committed:
Merge pull request #30768 from caktus/QA-3629--final-tweaks

Refs QA-3629 some final tweaks to case importer validation

orangejenny committed:
Combined migrations

orangejenny committed:
Updated casing

orangejenny commented:
All of this, except the final two cleanup commits, has been reviewed in other PRs.

AddisonDunn reviewed (approved):
Obviously, I only reviewed the last 2 commits, which look like simple enough changes to me.

kmtracey committed:
Fix SuggestedFieldTest to not not need DB access.

orangejenny committed:
Merge pull request #30771 from caktus/fix-suggested-fields-tests

Fix SuggestedFieldTest to not not need DB access.

snopoke commented on corehq/apps/data_dictionary/tests/utils.py:
```diff
@@ -1,10 +1,14 @@
         prop.data_type = data_type
         prop.save()
+        if prop_name in allowed_values:
+            for value in allowed_values[prop_name]:
+                CasePropertyAllowedValue.objects.create(case_property=prop, allowed_value=value)
```
can this be done in bulk? Also should this be a `get_or_create`?

orangejenny replied:
yeah, `get_or_create` makes sense. I could go either way on the bulk save, as there aren't likely to be that many values. @kmtracey could you update this?

orangejenny replied:
Updated in https://github.com/dimagi/commcare-hq/pull/30786

kmtracey committed:
Only narrow the width of "Create new property" column...

...when the validation feature flag is enabled. (Without the extra columns from the flag, the narrow width makes the remaining columns space out too much.)

Also fix a test utility to use get_or_create instead of simply create.

orangejenny committed:
Merge pull request #30786 from caktus/QA-3629--final-tweaks

Only narrow the width of "Create new property" column...

kmtracey committed:
Update messaging around data dictionary valid values...

...depending on validaton toggle setting.

orangejenny committed:
Merge pull request #30814 from caktus/data-dictionary-update-messaging

Update messaging around data dictionary valid values...

kmtracey committed:
Update data dictionary export/import...

To export the "display" value for the data types, and accept the same.

This makes editing of the data dictionary via web UI have the same names for data types as the export/import.

orangejenny committed:
Merge pull request #30849 from caktus/dd-validation-export-import-display-values

Update data dictionary export/import...

orangejenny committed:
Merge branch 'master' into jls/caktus-data-dictionary-release-2

orangejenny committed:
Updated migrations.lock

orangejenny committed:
Updated migrations.lock
